perf(api): allow to use an arena for return values

bfredl opened the pull request:
complement/alternative to #19877 . While the lua special case for buffer text is still worth it, for other API functions we should eliminate return value allocation without special casing lua-api at all.

if an API function takes an `Arena *arena` it is expected to use this for allocate return values. it is also admissible to use pointers to static memory or pointers owned elsewhere, like in the `nvim_buf_get_name` example. Any callsite of api-functions must consume returned memory immediately before invoking anything which might change editor state (and thus such pointers). Arenas are now also lazily initialized, so there are no xmalloc/xfree calls involved if the arena actually is not used.

bfredl committed:
perf(api): allow to use an arena for return values

bfredl committed:
refactor(arena): use a shared block freelist

This is both simpler in client code and more effective (always reuse
block hottest in cache)

lewis6991 commented on src/nvim/api/buffer.c:
```diff
@@ -1021,7 +1021,7 @@ void nvim_buf_del_var(Buffer buffer, String name, Error *err)
 /// @param[out] err   Error details, if any
 /// @return Buffer name
-String nvim_buf_get_name(Buffer buffer, Error *err)
+String nvim_buf_get_name(Buffer buffer, Arena *arena, Error *err)
```
Is the arena argument required here?

bfredl replied:
Yes, otherwise, the wrapper would have attempted to free the return value.

lewis6991 replied:
Ah yes, sorry of course.

bfredl committed:
refactor(api): provide a temporary copy solution for nvim_call_atomic

Make the copy_object() family accept an optional arena. More than
half of the callsites should be refactored to use an arena later
anyway.

lewis6991 commented:
Very nice!

I guess to optimize `nvim_buf_get_lines` we will still need something like #19877, so functions can have access to `lstate`. That is unless you have another magical idea on how not to special case Lua.

bfredl commented:
Let's still do #19877 for now, `nvim_buf_get_lines` is used more than frequently enough from lua to warrant its own micro-optimization.
